lib/keystore.js

```diff
@@ -0,0 +1,66 @@
+const Dexie = require('dexie').default;
+
+/** @description A keystore based on indexedDB */
+class Keystore {
```

Comment:
I think the current `Keystore` is a super thin wrapper of `dixie`  and it's not doing much work.
So, instead,  why don't we make this class more Scalar DL specific ?
More specifically, this constructor will accept `ClientService` and all the IndexedDB operations are done in this class.
(roughly speaking, moving `useIndexedDb` and `removeCachedPrivateKey` to this class. They should be renamed probably as well)
So, `ClientService` internally delegates only write/update operations (as done in useIndexedDB) for IndexedDB to this class and can return the class so that users can do further operations like removing keys or getting the contents.
Otherwise, `ClientService` will require more APIs as we have more IndexedDB operations.

As a class design, it's more isolated and more following to SRP.
(so, ClientService shouldn't be the one who manages IndexedDB operations, `KeyStore` should the one)
BTW, I feel `Keystore` should be renamed to `IndexedDb` for ease of understanding.

What do you think ? @supl  @KenYurino

Reply:
@feeblefakie 
I understand your concerns.

In fact, the indexedDB operations only rely on client properties.
How about we decorate the `properties` object.

```
const properties = await useIndexedDB({
    'scalar.dl.client.cert_holder_id': '...',
    'scalar.dl.client.cert_version': '...',
});
const clientService(properties);
```
and we execute the flow wtih properties only.
ClientService will be not aware of any about indexedDB.

Reply:
@scalarindetail 
I think it's better but I feel we can make it even better.
How about having two constructors like below ?
```
properties = .... // properties creation
new ClientService(properties);
```
and
```
properties = .... // properties creation
indexedDb = new IndexedDb(properties);
new ClientService(properties, indexedDb);
```

So that we don't have to manage the `enabled` flag ?

`ClientService` object will delegate for writing keys internally but will not delegate other operations. The other operations are done directly by IndexedDb object.

Reply:
Understood.

I have another idea.
Since ClientService itself doesn't really need to decide to write, read or delete keys
How about this.
```
const clientService = await withIndexedDB(new ClientService(properties));

// only if the application needs to remove the key
await clearIndexedDB(clientService);
```

Reply:
@scalarindetail That is also good. But the interface looks a bit tricky to me.
In your example, `withIndexedDb` is kind of a static factory ?
I would name them and construct them in a bit differently way like this.

```
// a bit long names for clarity
clientServiceWithIndexedDb = await new ClientServiceWithIndexedDb(new ClientService(properties));
// clientServiceWithIndexedDb will delegate all the operations of `ClientService` to ClientService object
// writing to the IndexedDb is done automatically by this object
// the following operations are done in the object as needed
clientServiceWithIndexedDb.getIndexedDb();
clientServiceWithIndexedDb.deleteIndexedDb();
```

Reply:
Now it is

```
let properties = {
    'scalar.dl.client.cert_holder_id': 'foo@example.com',
    'scalar.dl.client.cert_version': 1,
    ...
}; // Not specify 'scalar.dl.client.private_key_pem' or 'scalar.dl.client.private_key_cryptokey'

let clientService;
try {
    // try to read private key from indexedDB
    clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
} catch (err) {
    properties['scalar.dl.client.private_key_pem'] = /* from some place */

    // getIndexedDb will store the private key this time.
    clientService = await new ClientServiceWithIndexedDb(new ClientService(properties));
}
```